File: api/src/main/java/dev/obscuria/fragmentum/content/util/signal/Signal1.java
package dev.obscuria.fragmentum.content.util.signal;

import org.apache.commons.lang3.NotImplementedException;

@SuppressWarnings("unused")
public final class Signal1<P1> extends Signal<Signal1.Listener<P1>>
{
    public void emit(P1 p1)
    {
        throw new NotImplementedException();
    }

    @FunctionalInterface
    public interface Listener<P1>
    {
        void consume(P1 p1);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal1.kt
package dev.obscuria.fragmentum.tools.signal

class Signal1<P1> : Signal<Signal1.Listener<P1>>() {

    fun emit(p1: P1) = emit { it.consume(p1) }

    fun interface Listener<P1> {

        fun consume(p1: P1)
    }
}